builder/replica_pipeline_test.cc

```diff
@@ -0,0 +1,1129 @@
+  }
+}
+
+// clang-format off
+static std::function<pipeline()> kMultipleOutputsReplica =
+// BEGIN define_replica_pipeline() output
+[]() -> ::slinky::pipeline {
+  node_context ctx;
+  auto in = buffer_expr::make(ctx, "in", sizeof(uint32_t), 3);
+  auto sum_x = buffer_expr::make(ctx, "sum_x", sizeof(uint32_t), 2);
+  auto y = var(ctx, "y");
+  auto z = var(ctx, "z");
+  auto _1 = variable::make(in->sym());
+  auto _2 = buffer_min(_1, 0);
+  auto _3 = buffer_max(_1, 0);
+  auto _4 = buffer_min(_1, 1);
+  auto _5 = buffer_max(_1, 1);
+  auto _replica_fn_6 = [=](const buffer<const void>& i0, const buffer<void>& o0, const buffer<void>& o1) -> index_t {
+    const buffer<const void>* ins[] = {&i0};
+    const buffer<void>* outs[] = {&o0, &o1};
+    const func::input fins[] = {{in, {{_2, _3}, {_4, _5}, point(z)}}};
+    const std::vector<var> fout_dims[] = {{y, z}, {z}};
+    return ::slinky::internal::replica_pipeline_handler(ins, outs, fins, fout_dims);
+  };
+  auto _7 = buffer_min(_1, 0);
+  auto _8 = buffer_max(_1, 0);
+  auto _9 = buffer_min(_1, 1);
+  auto _10 = buffer_max(_1, 1);
+  auto sum_xy = buffer_expr::make(ctx, "sum_xy", sizeof(uint32_t), 1);
+  auto _fn_0 = func::make(std::move(_replica_fn_6), {{in, {{_7, _8}, {_9, _10}, point(z)}}}, {{sum_x, {y, z}}, {sum_xy, {z}}});
+  _fn_0.loops({{z, 1, loop_mode::serial}});
+  auto p = build_pipeline(ctx, {}, {in}, {sum_x, sum_xy}, {});
+  return p;
+}
+// END define_replica_pipeline() output
+;
+// clang-format on
+
+TEST_F(ReplicaPipelineTest, multiple_outputs) {
+  constexpr int split = 1;
+  constexpr loop_mode lm = loop_mode::serial;
+  // Make the pipeline
+  node_context ctx;
+
+  auto in = buffer_expr::make(ctx, "in", sizeof(int), 3);
+  auto sum_x = buffer_expr::make(ctx, "sum_x", sizeof(int), 2);
+  auto sum_xy = buffer_expr::make(ctx, "sum_xy", sizeof(int), 1);
+
+  var x(ctx, "x");
+  var y(ctx, "y");
+  var z(ctx, "z");
+
+  auto X = in->dim(0).bounds;
+  auto Y = in->dim(1).bounds;
+
+  // For a 3D input in(x, y, z), compute sum_x = sum(input(:, y, z)) and sum_xy = sum(input(:, :, z)) in one stage.
+  func::callable<const int, int, int> sum_x_xy = [](const buffer<const int>& in, const buffer<int>& sum_x,
+                                                     const buffer<int>& sum_xy) -> index_t {
+    assert(sum_x.dim(1).min() == sum_xy.dim(0).min());
+    for (index_t z = sum_xy.dim(0).min(); z <= sum_xy.dim(0).max(); ++z) {
+      sum_xy(z) = 0;
+      for (index_t y = sum_x.dim(0).min(); y <= sum_x.dim(0).max(); ++y) {
+        sum_x(y, z) = 0;
+        for (index_t x = in.dim(0).min(); x <= in.dim(0).max(); ++x) {
+          sum_x(y, z) += in(x, y, z);
+          sum_xy(z) += in(x, y, z);
+        }
+      }
+    }
+    return 0;
+  };
+  func sums = func::make(std::move(sum_x_xy), {{in, {X, Y, point(z)}}}, {{sum_x, {y, z}}, {sum_xy, {z}}});
+
+  if (split > 0) {
+    sums.loops({{z, split, lm}});
+  }
+
+  pipeline p = build_pipeline(ctx, {in}, {sum_x, sum_xy});
```

Comment:
Saving this one for last since it's nontrivial surgery... so, to be clear, what code would remain in `replica_pipeline_test.cc`? The generated replicas in their current form, but no standalone tests at all? (i.e. the generated replica fns would be called by the tests in pipeline_test?)

We could certainly do that, and it would reduce duplication, but before we do that, I have a slight concern about this; consider that pipeline_test.cc checks all its pipelines for correctness; with this change, the tests in `pipeline_test.cc` now have multiple purposes, including having some (but not all) of its test calling and running a replica pipeline, but not checking them for correctness. I'd kinda rather have pipeline_test.cc remain 'pure' in that it validates that lots of pipeline variants produce correct results, and nothing else.

Reply:
I think it *does* check for everything the current test does, just indirectly: because the output must match the golden file, and the golden file is itself a test that runs.

> I'd kinda rather have pipeline_test.cc remain 'pure' in that it validates that lots of pipeline variants produce correct results, and nothing else.

This is way too much duplication:
- We're going to be changing the func/buffer_expr API in significant ways I'm sure. This duplication means a significant extra effort in that regard.
- The pipeline tests change often as we find and fix bugs, letting them diverge isn't the worst thing, but it isn't great either. New capabilities are unlikely to be added to the replica tests.

Also: I'm already doing almost exactly this with the visualizer tests. pipeline_test.cc runs the visualizer and saves the result (and does nothing else with it). These replica tests are strictly better than that because we have some reasonable golden output to compare against.

I'm familiar with the "tests should only test one thing" argument, but TBH, I almost always find that more hassle than it's worth. If we stuck to that rigidly, what we'd need to do here is refactor the pipeline definitions into helper functions shared by multiple tests. But for what? I don't see what this rule gives us, and it makes other things a pain (we'd need to expose a lot of internal pipeline definition state for testing purposes).

Reply:
OK, I'll revise accordingly.

Reply:
Implemented according to what I think you want. PTAL.